File: 086_Minimum Absolute Difference in BST.cpp
// Minimum Absolute Difference in BST
// Find min absolute difference between values of any two nodes in a BST.
#include <bits/stdc++.h>
using namespace std;
struct TreeNode{int val; TreeNode* left; TreeNode* right; TreeNode(int x): val(x), left(nullptr), right(nullptr) {}};

int getMinimumDifference(TreeNode* root){
    int prev = INT_MIN; bool hasPrev=false; int best = INT_MAX;
    function<void(TreeNode*)> dfs = [&](TreeNode* node){
        if(!node) return;
        dfs(node->left);
        if(hasPrev) best = min(best, abs(node->val - prev)); else hasPrev = true;
        prev = node->val;
        dfs(node->right);
    };
    dfs(root);
    return best==INT_MAX?0:best;
}

int main(){
    TreeNode* root = new TreeNode(1);
    root->right = new TreeNode(3);
    root->right->left = new TreeNode(2);
    cout<<"Min abs diff: "<<getMinimumDifference(root)<<"\n"; // 1
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(h)

File: 086_Minimum Absolute Difference in BST.py
"""
Minimum Absolute Difference in BST
Find minimum absolute difference between values of any two nodes in a BST.
"""
from typing import Optional

class TreeNode:
    def __init__(self,val=0,left=None,right=None):
        self.val=val; self.left=left; self.right=right

def get_minimum_difference(root: Optional[TreeNode]) -> int:
    prev = None; best = float('inf')
    def dfs(node):
        nonlocal prev, best
        if not node: return
        dfs(node.left)
        if prev is not None:
            best = min(best, node.val - prev)
        prev = node.val
        dfs(node.right)
    dfs(root)
    return int(best) if best!=float('inf') else 0

if __name__ == '__main__':
    root = TreeNode(1)
    root.right = TreeNode(3)
    root.right.left = TreeNode(2)
    print('Min abs diff:', get_minimum_difference(root))

# Time Complexity: O(n)
# Space Complexity: O(h)
